lib/srv/db/mysql/protocol/packet.go

```diff
@@ -108,103 +53,13 @@ func ParsePacket(conn io.Reader) (Packet, error) {
-		if len(packetBytes) < minLen {
-			return nil, trace.BadParameter("failed to parse ERR packet: %v", packetBytes)
-		}
-		return &Error{packet: packet, message: string(packetBytes[minLen:])}, nil
-
-	case mysql.COM_QUERY:
-		// Be a bit paranoid and make sure the packet is not truncated.
-		if len(packetBytes) < packetHeaderAndTypeSize {
-			return nil, trace.BadParameter("failed to parse COM_QUERY packet: %v", packetBytes)
-		}
-		// 4-byte packet header + 1-byte payload header, then query text.
-		return &Query{packet: packet, query: string(packetBytes[packetHeaderAndTypeSize:])}, nil
-
-	case mysql.COM_QUIT:
-		return &Quit{packet: packet}, nil
-
-	case mysql.COM_CHANGE_USER:
-		if len(packetBytes) < packetHeaderAndTypeSize {
-			return nil, trace.BadParameter("failed to parse COM_CHANGE_USER packet: %v", packetBytes)
-		}
-		// User is the first null-terminated string in the payload:
-		// https://dev.mysql.com/doc/internals/en/com-change-user.html#packet-COM_CHANGE_USER
-		idx := bytes.IndexByte(packetBytes[packetHeaderAndTypeSize:], 0x00)
-		if idx < 0 {
-			return nil, trace.BadParameter("failed to parse COM_CHANGE_USER packet: %v", packetBytes)
-		}
-		return &ChangeUser{packet: packet, user: string(packetBytes[packetHeaderAndTypeSize : packetHeaderAndTypeSize+idx])}, nil
-
-	case mysql.COM_STMT_PREPARE:
-		packet, ok := parseStatementPreparePacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_PREPARE packet: %v", packetBytes)
-		}
-		return packet, nil
-
-	case mysql.COM_STMT_SEND_LONG_DATA:
-		packet, ok := parseStatementSendLongDataPacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_SEND_LONG_DATA packet: %v", packetBytes)
-		}
-		return packet, nil
-
-	case mysql.COM_STMT_EXECUTE:
-		packet, ok := parseStatementExecutePacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_EXECUTE packet: %v", packetBytes)
-		}
-		return packet, nil
-
-	case mysql.COM_STMT_CLOSE:
-		packet, ok := parseStatementClosePacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_CLOSE packet: %v", packetBytes)
-		}
-		return packet, nil
-
-	case mysql.COM_STMT_RESET:
-		packet, ok := parseStatementResetPacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_RESET packet: %v", packetBytes)
-		}
-		return packet, nil
-
-	case mysql.COM_STMT_FETCH:
-		packet, ok := parseStatementFetchPacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_FETCH packet: %v", packetBytes)
-		}
-		return packet, nil
-
-	case packetTypeStatementBulkExecute:
-		packet, ok := parseStatementBulkExecutePacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_BULK_EXECUTE packet: %v", packetBytes)
-		}
-		return packet, nil
+	if int(packetType) < len(packetParsersByType) && packetParsersByType[packetType] != nil {
```

Comment:
This is an interesting approach to "use a slice to replace the switch for tiny performance improvement" (quoting the PR description). Naively I'd expect the switch to have better performance. Are the benefits of this approach discussed anywhere?

Reply:
my believe is switch is `o(n)` but slice is `o(1)`. i couldn't a good post ppl comparing slice vs switch. switch is mostly compared with map. here is one stackoverflow has some benchmark https://stackoverflow.com/questions/46789259/map-vs-switch-performance-in-go. Though I believe switch will get slower when n gets large.

I will do some benchmark code for fun and share

Reply:
some rough numbers
```
go test -bench=. -benchtime=10s

goos: darwin
goarch: arm64
pkg: github.com/greedy52/golang-switch-slice-map-compare/intkey
BenchmarkBaseline-8   	860302086	        13.52 ns/op
BenchmarkSwitch1-8    	888747649	        13.50 ns/op
BenchmarkSwitch5-8    	828560742	        14.49 ns/op
BenchmarkSwitch10-8   	489604197	        24.49 ns/op
BenchmarkSwitch20-8   	398235208	        30.13 ns/op
BenchmarkSlice-8      	887899017	        13.49 ns/op
BenchmarkMap-8        	339221132	        35.84 ns/op
```

https://github.com/greedy52/golang-switch-slice-map-compare/blob/main/intkey/benchmark_test.go